### ISSUE TITLE
training, validation and test split

### ISSUE THREAD

**Issue by paulhoule** [minimaxir/textgenrnn#20]
Can't really evaluate or improve without it.

**Comment by minimaxir** [comment]
textgenrnn is a rare statistical case where overfitting is the *desired* behavior, and the results (predicted text) are more qualitative than quantitative.

**Comment by paulhoule** [comment]
Judged quantitatively the textgenrnn results currently suck.

I have done experiments with similar models.  (Generally plain stacked LSTM or GRU)  The main difference is that

I've found that severe overfitting does not produce good results.  I've found that validation loss is predictive of "will I like the results?"

**Comment by minimaxir** [comment]
Reopening to add for v1.2:

* Validation training if prop_keep < 1.0
* `SpatialDropout1D`, which *may* make it more typo resilient and improve validation (it didn't in my testing and just made the model not converge).

**Comment by minimaxir** [comment]
Added both changes in v1.2

That's the extent I can do to mitigate overfitting.
